go-controller/pkg/util/gateway-init.go

```diff
@@ -139,61 +106,62 @@ func GatewayInit(clusterIPSubnet []string, nodeName, nicIP, physicalInterface,
+	routerMac, routerIP, err := GetPortAddresses(gwSwitchPort, false, false)
+	if routerMac == "" || routerIP == "" {
+		// Create the gateway switch port in 'join' if it doesn't exist yet
+		stdout, stderr, err = RunOVNNbctl("--wait=sb", "--may-exist",
+			"lsp-add", "join", gwSwitchPort,
+			"--", "--if-exists", "clear", "logical_switch_port",
+			gwSwitchPort, "dynamic_addresses",
+			"--", "lsp-set-addresses", gwSwitchPort, "dynamic")
+		if err != nil {
+			return fmt.Errorf("failed to add logical switch port %s, stdout: %q, "+
+				"stderr: %q, error: %v", gwSwitchPort, stdout, stderr, err)
+		}
+		// Wait for its addresses
+		routerMac, routerIP, err = GetPortAddresses(gwSwitchPort, false, true)
+		if err != nil {
+			return fmt.Errorf("Error while waiting for addresses for gateway switch port %q: %v",
+				gwSwitchPort, err)
+		}
 	}
 
-	var routerIP string
-	if routerMac == "" {
-		routerMac = GenerateMac()
-		if err = func() error {
-			err = lockNBForGateways()
-			if err != nil {
-				return err
-			}
-			defer unlockNBForGateways()
-			routerIP, err = generateGatewayIP()
-			if err != nil {
-				return err
-			}
+	// Grab the 'join' switch prefix to add to our gateway router's IP
+	joinSubnet, stderr, err := RunOVNNbctl("--if-exists", "get",
+		"logical_switch", "join", "other-config:subnet")
+	if err != nil {
+		return fmt.Errorf("Failed to get 'join' switch subnet:  %s, stderr: %q, %v",
+			joinSubnet, stderr, err)
+	}
+	if joinSubnet == "" {
+		return fmt.Errorf("Empty 'join' switch subnet")
+	}
+	joinSubnetSplit := strings.Split(strings.Trim(joinSubnet, `"`), "/")
+	if len(joinSubnetSplit) != 2 {
+		return fmt.Errorf("Invalid 'join' switch subnet %q", joinSubnet)
+	}
+	routerIPCIDR := routerIP + "/" + joinSubnetSplit[1]
 
-			stdout, stderr, err = RunOVNNbctl("--", "--may-exist", "lrp-add",
-				gatewayRouter, "rtoj-"+gatewayRouter, routerMac, routerIP,
-				"--", "set", "logical_router_port", "rtoj-"+gatewayRouter,
-				"external_ids:connect_to_join=yes")
-			if err != nil {
-				return fmt.Errorf("failed to add logical port to router, stdout: %q, "+
-					"stderr: %q, error: %v", stdout, stderr, err)
-			}
-			return nil
-		}(); err != nil {
-			return err
-		}
+	stdout, stderr, err = RunOVNNbctl("--", "--may-exist", "lrp-add",
+		gatewayRouter, gwRouterPort, routerMac, routerIPCIDR)
+	if err != nil {
+		return fmt.Errorf("failed to add logical port to router, stdout: %q, "+
+			"stderr: %q, error: %v", stdout, stderr, err)
 	}
 
-	if routerIP == "" {
-		stdout, stderr, err = RunOVNNbctl("--if-exists", "get",
-			"logical_router_port", "rtoj-"+gatewayRouter, "networks")
-		if err != nil {
-			return fmt.Errorf("failed to get routerIP for %s "+
-				"stdout: %q, stderr: %q, error: %v",
-				"rtoj-"+gatewayRouter, stdout, stderr, err)
-		}
-		routerIP = strings.Trim(stdout, "[]\"")
+	stdout, stderr, err = RunOVNNbctl("set", "logical_switch_port",
+		gwSwitchPort, "type=router", "options:router-port="+gwRouterPort)
+	if err != nil {
```

Comment:
I just had a cursory glance at OVN's IPAM code. It has the following piece.
```

struct ovn_port *op = ovn_port_find(ports, nbsp->name);
            if (!op || op->nbsp != nbsp || op->peer) {
                /* Do not allocate addresses for logical switch ports that
                 * have a peer. */
                continue;
            }
```
So I think, in your case it likely worked because, you allocated the IP before setting options:router-port. Mark Michelson (@putnopvut) , who has recently enhanced IPAM quite a bit, may understand the implications here better.

Reply:
Hi guys.

Guru is 100% correct. ovn-northd will not allocate dynamic addresses to a logical switch port of type "router" that is connected to a router port. ovn-northd purposely never assigns the first address in a subnet to any switch ports. The intention seems to be to allow for you to reliably set that address on the router switch port yourself.

Of course, if you assign the dynamic address prior to setting the router-port, then ovn-northd will assign an address. The fact that this address stays assigned after the port type and router-port are specified seems buggy to me. I think that ovn-northd should remove the dynamic address from the port.

If ovn-northd were to allow dynamic addresses to be assigned to router switch ports, then that would require an enhancement to be made to ovn-northd's IPAM. As you've pointed out, a switch can be attached to multiple routers, so simply assuming to assign the first IP address in the subnet is not good enough. Presumably, the easiest path would be to treat these ports no differently than other logical switch ports. It may be worth asking upstream to find out why the current behavior was written the way it was, just in case there is something more subtle I'm not thinking about.

Reply:
>It may be worth asking upstream to find out why the current behavior was written the way it was, just in case there is something more subtle I'm not thinking about.

I guided an intern to initially add the ipam feature. Because a switch with a "peer" would not be a logical port like a VM, we wanted to skip assigning any IP to it. Because the "peer" itself will have its own IP (assigned independently).

Reply:
@putnopvut If we create a lsp with addresses as "dynamic", get its dynamic_address and then set it to static address column (clearing out dynamic_addresses), and then later set the options:router-port, we would be okay, right?

Reply:
> edit: On second thoughts, that will likely set us up for race conditions, if another node sets "dynamic" on the same switch at the same time.

@shettyg can you elaborate on where you think the race would be here?

Reply:
@dcbw 

Long story, but the summary is that after reading the IPAM code again, I think we are okay.

Reply:
As long as you make the changes to the port as a single transaction, I'm not sure that race conditions could arise.

If you
- read the current dynamic addresses
- set the static address
- and clear the dynamic addresses
in a single transaction, then you should have well-define "happens before" behavior.

Reply:
@putnopvut @shettyg its sounding like this is going to be messy.  I'm happy to drop the PR since it seems like OVN isn't expecting this.  I was just reviewing the code and wondered why we couldn't just get OVN to dynamically allocate the IPs but didn't understand the implications for OVN.

Or should I keep this one going?

Reply:
I think the current central locking mechanism is poor. I think this is an improvement.

Reply:
@shettyg @putnopvut updated; but it turns out if I clear dynamic addresses, even after setting the LSP as type=router I get IP conflicts. Could you double-check my logic in the updated PR?